src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,705 @@
+    void (*validXLogRecord)(const XLogRecord *record, PgPageSize heapPageSize);
+    pg_crc32 (*recordChecksum)(const XLogRecord *record, PgPageSize heapPageSize);
+} WalInterface;
+
+static WalInterface interfaces[] = {
+    {
+        PG_VERSION_94,
+        CFGOPTVAL_FORK_GPDB,
+        GPDB6_XLOG_PAGE_MAGIC,
+        validXLogRecordHeaderGPDB6,
+        validXLogRecordGPDB6,
+        recordChecksumGPDB6
+    }
+};
+
+typedef struct WalFilter
+{
+    ReadStep currentStep;
+    const unsigned char *page;
+
+    PgPageSize heapPageSize;
+    PgPageSize walPageSize;
+    uint32_t segSize;
+
+    bool isBegin;
+
+    size_t pageOffset;
+    size_t inputOffset;
+    XLogRecPtr recPtr;
+
+    XLogPageHeaderData *currentHeader;
+
+    XLogRecord *record;
+    uint32_t recBufSize;
+    // Offset to the body of the record on the current page
+    size_t headerOffset;
+    // How many bytes we read of this record
+    size_t gotLen;
+    // Total size of record on current page
+    size_t totLen;
+
+    List *headers;
+
+    WalInterface *walInterface;
+
+    const ArchiveGetFile *archiveInfo;
+
+    // Records count for debug
+    uint32_t recordNum;
+
+    bool done;
+    bool sameInput;
+    bool isSwitchWal;
+} WalFilterState;
+
+/***********************************************************************************************************************************
+Render as string for logging
+***********************************************************************************************************************************/
+static void
+walFilterToLog(const WalFilterState *const this, StringStatic *const debugLog)
+{
+    strStcFmt(
+        debugLog,
+        "{recordNum: %u, step: %u isBegin: %s, pageOffset: %zu, inputOffset: %zu, recBufSize: %u, gotLen: %zu, totLen: %zu}",
+        this->recordNum,
+        this->currentStep,
+        this->isBegin ? "true" : "false",
+        this->pageOffset,
+        this->inputOffset,
+        this->recBufSize,
+        this->gotLen,
+        this->totLen
+        );
+}
+
+#define FUNCTION_LOG_WAL_FILTER_TYPE                                                                                               \
+    ZstDecompress *
+#define FUNCTION_LOG_WAL_FILTER_FORMAT(value, buffer, bufferSize)                                                                  \
+    FUNCTION_LOG_OBJECT_FORMAT(value, walFilterToLog, buffer, bufferSize)
+
+static inline
+void
+checkOutputSize(Buffer *const output, const size_t size)
+{
+    if (bufRemains(output) <= size)
+    {
+        bufResize(output, bufUsed(output) + size);
+    }
+}
+
+// Reading the next page from the input buffer. If the input buffer is exhausted, remember the current step and returns false.
+// In this case, we should exit the process function to get a new input buffer.
+// Returns true on success page read.
+static inline
+bool
+readPage(WalFilterState *const this, const Buffer *const input, const ReadStep step)
```

Comment:
Maybe we should rename this function to findPage, because in fact it doesn't read anything, but rather finds the page?